Enhance coreclr processing of event pipe configuration to be able to write the pid of the current process into the nettrace file

davidwrighton opened the pull request:
- Also fix the behavior of get_next_config_value_as_utf8_string to work correctly on coreclr, it was type-punning between a buffer returned by ep_rt_byte_array_alloc and ep_rt_utf8_string_free, which isn't safe in the coreclr implementation

davidwrighton committed:
Enhance coreclr processing of event pipe configuration to be able to write the pid of the current process into the nettrace file
- Also fix the behavior of get_next_config_value_as_utf8_string to work correctly on coreclr, it was type-punning between a buffer returned by ep_rt_byte_array_alloc and ep_rt_utf8_string_free, which isn't safe in the coreclr implementation

lateralusX reviewed (commented):
Initially we didn't have a need for a pure UTF8 string alloc function and didn't add it, mainly since it caused some confusion around byte vs number of characters as the argument when calling alloc and we tried to stay away as much as possible regarding length of UTF8 strings (there is no ep_rt_utf8_string_len for example) and try to fallback to using len -1 when doing conversions between UTF8 and UTF16.

We have a ep_rt_utf8_string_dup function for null terminated strings maybe an alternative could be to add a range based dup function for UTF8 strings (start/end) that could be used in get_next_config_value_as_utf8_string, pretty much doing ptrdiff + 1 using runtimes string allocator (malloc for example) and then doing a memcpy and NULL termination if needed).

lateralusX commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:
```diff
@@ -1569,7 +1569,34 @@ ep_rt_config_value_get_output_path (void)
+	{
+		while (true)
+		{
+			const char* pidSearchString = "{pid}";
```
nit: should we use same string type through function (const ep_char8_t *)?

lateralusX commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:
```diff
@@ -1569,7 +1569,34 @@ ep_rt_config_value_get_output_path (void)
+				size_t newBufSize = strlen(outputPath) + 12; // 12 is 1 for the null-terminator + 11 for the largest possible 4 byte integer converted to a string
+				*outputPathPid = '\0';
+				ep_char8_t *newOutputPath =  reinterpret_cast<ep_char8_t *>(malloc(newBufSize));
+				sprintf_s(newOutputPath, newBufSize, "%s%d%s", outputPath, (int)GetCurrentProcessId(), outputPathPid + strlen(pidSearchString));
```
check for malloc failure.

lateralusX commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:
```diff
@@ -1569,7 +1569,34 @@ ep_rt_config_value_get_output_path (void)
+		}
+		return outputPath;
+	}
+	else
```
nit: no need for else, just return outputPath (since it will be NULL when not entering if statement).

josalem commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:
```diff
@@ -1569,7 +1569,34 @@ ep_rt_config_value_get_output_path (void)
+			ep_char8_t* outputPathPid = strstr(outputPath, pidSearchString);
+			if (outputPathPid != NULL)
+			{
+				size_t newBufSize = strlen(outputPath) + 12; // 12 is 1 for the null-terminator + 11 for the largest possible 4 byte integer converted to a string
```
If I recall, macOS uses 64-bit PIDs.  Does `GetCurrentProcessId` account for that or does it always truncate to a 32-bit value?

davidwrighton replied:
@janvorli Do I need to worry about 64bit pids? I was assuming that the PAL would insulate me here, but if some supported OS might actually use a 64bit pid one could get in trouble. Do you have any data here?

josalem reviewed (approved):
This seems like a good addition to this activation mechanism.  This change should also be documented [here](https://github.com/dotnet/docs/blob/master/docs/core/diagnostics/eventpipe.md) before it ships in a preview.

Could you add a CI test for the PID insertion to make sure it's working across all supported platforms?

CC dotnet/diagnostics#515

davidwrighton committed:
Address feedback

davidwrighton commented:
@lateralusX  I've addressed your feedback, I like the range based dup.

@josalem, Do we have any existing CI tests which test this sort of environment variable in our current CI that you are aware of? I'd like to follow a pattern if we've done that somewhere.

lateralusX commented:
One though, shouldn't this feature be available in Mono as well, looks like it is currently implemented as part of CoreClr's runtime shim, meaning it won't work for Mono.

davidwrighton committed:
Update to make {pid} behavior work for mono as well and add a test

josalem commented on src/tests/tracing/eventpipe/complus_config/name_config_with_pid.cs:
```diff
@@ -0,0 +1,63 @@
+            {
+                Console.WriteLine($"{expectedPath} found");
+                return 100; 
+            }
```
We should wrap this in a try/finally that attempts to delete the file it potentially created.

josalem commented:
> @josalem, Do we have any existing CI tests which test this sort of environment variable in our current CI that you are aware of? I'd like to follow a pattern if we've done that somewhere.

You could use the code in the `src/tests/tracing/eventpipe/common` directory, but I think what you wrote up is sufficient.

davidwrighton committed:
Fix unix printf validation error

lateralusX commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:
```diff
@@ -2472,6 +2487,38 @@ ep_rt_utf8_string_strtok (
+			*str = NULL;
+			return false;
+		}
+		ep_rt_utf8_string_snprintf(newStr, newStrSize, "%.*s%s%s", (int)(strFound - (*str)), *str, strReplacement, strFound + strSearchLen);
```
I would probably have done this using memcpy, its faster (lowered to intrinsic by most compilers) and no risk of having any interpretation done regarding code points (since it's UTF8) in underlying implementation of snprintf.

davidwrighton replied:
True, memcpy would be a fair bit faster, but this should compile down to a bit less code as all of these strings have variable length, and in addition this is a rare path that is already quite slow such that the perf difference of using printf shouldn't be significant. I don't plan to change this.

lateralusX replied:
OK, was thinking that we added a generic replace function and that it might get called from other places in future that is not in a rare/slow code path, but guess we could optimize this later when we hit such case (and it turns out to be a problem). Just tried a very small benchmark using memcpy instead of snprintf for the above function and memcpy version was ~2.5x faster on msvc (for a ~70 byte string replacing {pid} with 123456789) timed over 1 million iterations.

davidwrighton committed:
Address test concerns
